### PR TITLE
Per API Per Gateway Feature

### PR DESCRIPTION
**Pull request by ChamNDeSilva**
### Proposed changes in this pull request

Adding per API per Gateway Feature

### When should this PR be merged

When we can update transport and msf4j versions for C5

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 1]
```diff
@@ -1597,6 +1600,85 @@ public String getEndpointConfig(String endpointId) throws APIMgtDAOException {
...
+                statement.setTimestamp(2, Timestamp.valueOf(LocalDateTime.now()));
+                statement.setString(3, dedicatedGateway.getUpdatedBy());
+                statement.setString(4, apiId);
+
+                // if the labels are null or perAPI-GW
+                if (labels == null || labels.isEmpty() ||
+                        labels.contains(ContainerBasedGatewayConstants.PER_API_GATEWAY_PREFIX + apiId)) {
+                    deleteLabelsMapping(connection, apiId);
+                    addLabelMapping(connection, apiId, labels);
+
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: do we need space here?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 2]
```diff
@@ -1597,6 +1600,85 @@ public String getEndpointConfig(String endpointId) throws APIMgtDAOException {
...
+                }
+                statement.execute();
+                connection.commit();
+            }  catch (SQLException e) {
+                String msg = "Couldn't update dedicated Gateway details of API : " + apiId;
+                connection.rollback();
+                log.error(msg, e);
+                throw new APIMgtDAOException(msg, ExceptionCodes.APIM_DAO_EXCEPTION);
+            } finally {
+                connection.setAutoCommit(DAOUtil.isAutoCommit());
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: what the purpose of this line?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 3]
```diff
@@ -1597,6 +1600,85 @@ public String getEndpointConfig(String endpointId) throws APIMgtDAOException {
...
+            }
+        } catch (SQLException e) {
+            throw new APIMgtDAOException("Error Executing query for updating Container Based Gateway",
+                    ExceptionCodes.APIM_DAO_EXCEPTION);
+
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: space?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java [hunk 4]
```diff
@@ -89,6 +95,8 @@ public void addCompositeAPI(CompositeAPI api) throws GatewayException {
         gatewayDTO.setApiSummary(apiSummary);
         publishToPublisherTopic(gatewayDTO);
 
+        // adding to the dedicated Gateway will happen as a seperate REST call
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java]: wht the purpose of this comment?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java [hunk 5]
```diff
@@ -114,6 +122,10 @@ public void deleteAPI(API api) throws GatewayException {
         if (log.isDebugEnabled()) {
             log.debug("API : " + api.getName() + " deleted event has been successfully published to broker");
         }
+        if (api.hasOwnGateway()) {
+            // Delete the Gateway - check how we can assume that we complete the deletion
+            removeContainerBasedGateway(api.getLabels().toArray()[0].toString(), api.getId());
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java]: possible NPE?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java [hunk 6]
```diff
@@ -421,6 +433,66 @@ public void deleteBlockCondition(BlockConditions blockConditions) throws Gateway
             }
         }
     }
+    /**
+     * @see APIGateway#createContainerBasedGateway(String, String)
+     */
+    @Override
+    public void createContainerBasedGateway(String apiId, String label) throws GatewayException {
+
+        if ("kubernetes".equalsIgnoreCase(config.getContainerGatewayConfigs().getCmsType())) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java]: lets uise constants for kurbunetees and openshift

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java [hunk 7]
```diff
@@ -421,6 +433,66 @@ public void deleteBlockCondition(BlockConditions blockConditions) throws Gateway
...
+
+    }
+
+    /**
+     * @see APIGateway#removeContainerBasedGateway(String, String)
+     */
+    @Override
+    public void removeContainerBasedGateway(String label, String apiId) throws GatewayException {
+
+        if ("kubernetes".equalsIgnoreCase(config.getContainerGatewayConfigs().getCmsType())) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java]: use constants

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java [hunk 8]
```diff
@@ -231,6 +233,65 @@ public API getAPIbyUUID(String uuid) throws APIManagementException {
...
+                    //add to the API
+                    labelSet.add(ContainerBasedGatewayConstants.PER_API_GATEWAY_PREFIX + apiId);
+                }
+            } else {
+                labelSet = getAPIbyUUID(apiId).getLabels();
+            }
+
+            // create or remove dedicated Gateway
+            gateway.updateDedicatedGateway(api, labelSet, dedicatedGateway.isEnabled());
+
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java]: space?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java [hunk 9]
```diff
@@ -231,6 +233,65 @@ public API getAPIbyUUID(String uuid) throws APIManagementException {
...
+            getApiDAO().updateDedicatedGateway(dedicatedGateway, apiId, labelSet);
+        } catch (APIMgtDAOException e) {
+            String errorMsg = "Error occurred while updating dedicatedGateway details of API with id " + apiId;
+            log.error(errorMsg, e);
+            throw new APIManagementException(errorMsg, e, e.getErrorHandler());
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java]: is this the correct way we agreed to thrw exceptions?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java [hunk 10]
```diff
@@ -231,6 +233,65 @@ public API getAPIbyUUID(String uuid) throws APIManagementException {
...
+    /**
+     * @see APIPublisherImpl#getDedicatedGateway(String)
+     */
+    @Override
+    public DedicatedGateway getDedicatedGateway(String apiId) throws APIManagementException {
+       DedicatedGateway dedicatedGateway;
+        try {
+            dedicatedGateway = getApiDAO().getDedicatedGateway(apiId);
+        } catch (APIMgtDAOException e) {
+            String errorMsg = "Error occurred while retrieving dedicated Gateway details of API with id " + apiId;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java]: is this the correct way we agreed to thrw exceptions?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 11]
```diff
@@ -1787,4 +1790,62 @@ public void removeObserver(EventObserver observer) {
...
+                    labelList.add(autoGenLabel);
+                    //Add to the db
+                    getLabelDAO().addLabels(labelList);
+                    //add to the API
+                    labelSet.add(ContainerBasedGatewayConstants.PER_API_GATEWAY_PREFIX + apiId);
+                }
+            } else {
+                labelSet = getAPIbyUUID(apiId).getLabels();
+            }
+
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: space?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 12]
```diff
@@ -1787,4 +1790,62 @@ public void removeObserver(EventObserver observer) {
...
+            // create or remove dedicated Gateway
+            gateway.updateDedicatedGateway(api, labelSet, dedicatedGateway.isEnabled());
+
+            getApiDAO().updateDedicatedGateway(dedicatedGateway, apiId, labelSet);
+        } catch (APIMgtDAOException e) {
+            String errorMsg = "Error occurred while updating dedicatedGateway details of API with id " + apiId;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: is this the correct way we agreed to thrw exceptions?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/KubernetesGatewayImpl.java [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+     * String serviceName, String namespace, String label)
+     */
+    @Override
+    public void createContainerBasedService(String serviceTemplate, String apiId, String serviceName, String namespace,
+                                            String label) throws GatewayException {
+
+        if (masterURL != null) {
+            List<HasMetadata> resources;
+
+            System.setProperty("kubernetes.auth.tryKubeConfig", "false");
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/KubernetesGatewayImpl.java]: do we need to override these properties? Should we read from config?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/KubernetesGatewayImpl.java [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+                            ExceptionCodes.NO_RESOURCE_LOADED_FROM_DEFINITION);
+                }
+                HasMetadata resource = resources.get(0);
+
+                if (resource instanceof Service) {
+                    // check whether there are existing service already
+                    if (client.services().inNamespace(namespace).withLabel(label) != null) {
+                        Service result = client.services().inNamespace(namespace).load(inputStream).create();
+                        if (log.isDebugEnabled()) {
+                            log.debug("The Kubernetes Service Definition :\n " + serviceTemplate);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/KubernetesGatewayImpl.java]: use single line debug statemetns

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/KubernetesGatewayImpl.java [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+    /**
+     * @see ContainerBasedGatewayGenerator#createContainerBasedDeployment(String deploymentTemplate, String apiId,
+     * String deploymentName, String namespace, String label)
+     */
+    @Override
+    public void createContainerBasedDeployment(String deploymentTemplate, String apiId, String deploymentName,
+                                               String namespace, String label) throws GatewayException {
+
+        if (masterURL != null) {
+            System.setProperty("kubernetes.auth.tryKubeConfig", "false");
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/KubernetesGatewayImpl.java]: will we need to ovverride these propertties?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/KubernetesGatewayImpl.java [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+    }
+
+    /**
+     * @see ContainerBasedGatewayGenerator#removeContainerBasedGateway(String label, String apiId, String namespace)
+     */
+    @Override
+    public void removeContainerBasedGateway(String label, String apiId, String namespace) throws GatewayException {
+
+        if (masterURL != null) {
+            System.setProperty("kubernetes.auth.tryKubeConfig", "false");
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/KubernetesGatewayImpl.java]: same here

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/OpenShiftGatewayImpl.java [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+    /**
+     * @see ContainerBasedGatewayGenerator#createContainerBasedService(String gatewayServiceTemplate, String apiId,
+     * String serviceName, String namespace, String label)
+     */
+    @Override
+    public void createContainerBasedService(String serviceTemplate, String apiId, String serviceName, String namespace,
+                                            String label) throws GatewayException {
+        if (masterURL != null) {
+
+            List<HasMetadata> resources;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/OpenShiftGatewayImpl.java]: same here configs should read from configs?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/OpenShiftGatewayImpl.java [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+    }
+
+    /**
+     * @see ContainerBasedGatewayGenerator#removeContainerBasedGateway(String label, String apiId, String namespace)
+     */
+    @Override
+    public void removeContainerBasedGateway(String label, String apiId, String namespace) throws GatewayException {
+
+        if (masterURL != null) {
+            System.setProperty("kubernetes.auth.tryKubeConfig", "false");
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/OpenShiftGatewayImpl.java]: cant we move client initlaization to single place? I see lots of places which do the same

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/ContainerBasedGatewayUtils.java [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    public static String getServiceAccountAccessToken() throws GatewayException {
+
+        //todo : get this read from the kube secret
+        String token;
+         File tokenFile = new File(saTokenFile);
+        try (InputStream inputStream = new FileInputStream(tokenFile);
+             InputStreamReader inputStreamReader = new InputStreamReader(inputStream, StandardCharsets.UTF_8);
+             BufferedReader bufferedReader = new BufferedReader(inputStreamReader)) {
+
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/ContainerBasedGatewayUtils.java]: spaces?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/ContainerBasedGatewayUtils.java [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Class to hold Util methods used in container based Gateway
+ */
+public class ContainerBasedGatewayUtils {
+
+    private static APIMConfigurations apimConfigurations = ServiceReferenceHolder.getInstance().getAPIMConfiguration();
+    //todo : temporary
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/ContainerBasedGatewayUtils.java]: temporary?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/workflow/APIStateChangeWorkflow.java [hunk 21]
```diff
@@ -133,10 +143,51 @@ public WorkflowResponse completeWorkflow(WorkflowExecutor workflowExecutor) thro
...
+
+            boolean hasOwnGateway =
+                    Boolean.valueOf(getAttribute(APIMgtConstants.WorkflowConstants.ATTRIBUTE_HAS_OWN_GATEWAY));
+            String apiId = getAttribute(APIMgtConstants.WorkflowConstants.ATTRIBUTE_API_NAME);
+            String label = getAttribute(APIMgtConstants.WorkflowConstants.ATTRIBUTE_API_AUTOGEN_LABEL);
+
+            if (hasOwnGateway) {
+                //assuming that there is no transition from (MAINTENANCE to  PROTOTYPED), (DEPRECATED to CREATED),
+                // (CREATED to DEPRECATED)
+                if ((currentState.equalsIgnoreCase("created") ||
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/workflow/APIStateChangeWorkflow.java]: cant we read states from constants?

**Comment by harsha89** [comment]
We need to add more debug logs and comments on the flow.

Also I do not see test cases. This willl severely effect code coverage as this is a very large pull requests. Please add test cases and do run builds locally and verify the ooverage

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java (hunk 5 above)
Comment (ChamNDeSilva) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java]: This is executed inside api.hasOwnGateway().
So having that value true means that the API has an autogenerated label always.
So will not give an NPE.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java (hunk 2 above)
Comment (ChamNDeSilva) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: It is added to the purpose of finally each transaction included in this method to be committed automatically when it is completed. Is it wrong ?

**Comment by ChamNDeSilva** [comment]
Closing the pull request for the moment. Will reopen in the future when resuming the implementation of this feature.

**Pull request closed without merging**